### PR TITLE
[Enhancement] Avoiding cache hot data being evicted due to a surge in cold data loading

### PR DESCRIPTION
**Pull request by sduzh**
Add property "datacache.partition_duration" for cloud native table. If a table is partitioned by DATE or DATETIME, and a partition range is within the last "datacache.partition_duration", it is a hot partition, queries on that partition will trigger cache population, otherwise will not.
"datacache.partition_duration" unit can be "second", "minute", "hour", "day", "week", "month", "year".

## What type of PR is this:
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes:

Fixes #

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/planner/OlapScanNode.java [hunk 1]
```diff
@@ -108,6 +109,16 @@ public class OlapScanNode extends ScanNode {
...
+    private final List<String> selectedPartitionNames = Lists.newArrayList();
+    private final List<Long> selectedPartitionVersions = Lists.newArrayList();
+    private final HashSet<Long> scanBackendIds = new HashSet<>();
+    // The column names applied dict optimization
+    // used for explain
+    private final List<String> appliedDictStringColumns = new ArrayList<>();
+    private final List<String> unUsedOutputStringColumns = new ArrayList<>();
+    // a bucket seq may map to many tablets, and each tablet has a TScanRangeLocations.
+    public ArrayListMultimap<Integer, TScanRangeLocations> bucketSeq2locations = ArrayListMultimap.create();
+    public List<Expr> prunedPartitionPredicates = Lists.newArrayList();
```
Comment (sduzh) [inline fe/fe-core/src/main/java/com/starrocks/planner/OlapScanNode.java:121]: Auto formatted(reordered), no logical change.

File: fe/fe-core/src/main/java/com/starrocks/planner/OlapScanNode.java [hunk 2]
```diff
@@ -133,35 +144,41 @@ public class OlapScanNode extends ScanNode {
...
-    public ArrayListMultimap<Integer, TScanRangeLocations> bucketSeq2locations = ArrayListMultimap.create();
-
     private List<Expr> bucketExprs = Lists.newArrayList();
     private List<ColumnRefOperator> bucketColumns = Lists.newArrayList();
-    public List<Expr> prunedPartitionPredicates = Lists.newArrayList();
-
     // record the selected partition with the selected tablets belong to it
     private Map<Long, List<Long>> partitionToScanTabletMap;
+    // The dict id int column ids to dict string column ids
+    private Map<Integer, Integer> dictStringIdToIntIds = Maps.newHashMap();
```
Comment (sduzh) [inline fe/fe-core/src/main/java/com/starrocks/planner/OlapScanNode.java:158]: reordered, no logical change.

File: fe/fe-core/src/main/java/com/starrocks/planner/OlapScanNode.java [hunk 3]
```diff
@@ -133,35 +144,41 @@ public class OlapScanNode extends ScanNode {
...
 
     // Constructs node to scan given data files of table 'tbl'.
     public OlapScanNode(PlanNodeId id, TupleDescriptor desc, String planNodeName) {
         super(id, desc, planNodeName);
         olapTable = (OlapTable) desc.getTable();
     }
 
+    // export some tablets
+    public static OlapScanNode createOlapScanNodeByLocation(
```
Comment (sduzh) [inline fe/fe-core/src/main/java/com/starrocks/planner/OlapScanNode.java:167]: reordered, no logical change.

File: fe/fe-core/src/main/java/com/starrocks/planner/OlapScanNode.java [hunk 4]
```diff
@@ -186,6 +203,10 @@ public List<Long> getSelectedPartitionIds() {
         return selectedPartitionIds;
     }
 
+    public void setSelectedPartitionIds(List<Long> selectedPartitionIds) {
```
Comment (sduzh) [inline fe/fe-core/src/main/java/com/starrocks/planner/OlapScanNode.java:206]: reordered, no logical change.

File: fe/fe-core/src/main/java/com/starrocks/catalog/PartitionKey.java [hunk 5]
```diff
@@ -132,6 +135,20 @@ public static PartitionKey createPartitionKey(List<PartitionValue> keys, List<Co
         return partitionKey;
     }
 
+    public static PartitionKey ofDateTime(LocalDateTime dateTime) throws AnalysisException {
```
Comment (sduzh) [inline fe/fe-core/src/main/java/com/starrocks/catalog/PartitionKey.java:138]: @Astralidea Please take a look at this file

File: fe/fe-core/src/main/java/com/starrocks/clone/DynamicPartitionScheduler.java [hunk 6]
```diff
@@ -503,10 +502,8 @@ private ArrayList<DropPartitionClause> getDropPartitionClauseByTTL(OlapTable ola
         List<Map.Entry<Long, Range<PartitionKey>>> candidatePartitionList = Lists.newArrayList();
 
         if (partitionType.isDateType()) {
-            LocalDateTime currentDateTime = LocalDateTime.now();
-            PartitionValue currentPartitionValue = new PartitionValue(currentDateTime.format(DateUtils.DATE_FORMATTER_UNIX));
-            PartitionKey currentPartitionKey = PartitionKey.createPartitionKey(
-                    ImmutableList.of(currentPartitionValue), partitionColumns);
+            PartitionKey currentPartitionKey = partitionType.isDatetime() ?
```
Comment (sduzh) [inline fe/fe-core/src/main/java/com/starrocks/clone/DynamicPartitionScheduler.java:505]: @Astralidea Please take a look at this file.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 119 / 173 (68.79%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/common/util/TimeUtils.java|0|9|00.00%|[343, 344, 345, 350, 351, 352, 353, 355, 356]|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|0|6|00.00%|[2027, 2028, 2030, 2031, 2032, 2033]|
|:large_blue_circle:|com/starrocks/catalog/RangePartitionInfo.java|0|1|00.00%|[563]|
|:large_blue_circle:|com/starrocks/sql/ast/PartitionValue.java|0|2|00.00%|[50, 54]|
|:large_blue_circle:|com/starrocks/lake/LakeTable.java|4|19|21.05%|[228, 229, 236, 237, 239, 240, 241, 242, 243, 244, 245, 246, 247, 248, 252]|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|3|7|42.86%|[2381, 2382, 2383, 2384]|
|:large_blue_circle:|com/starrocks/common/util/PropertyAnalyzer.java|3|5|60.00%|[967, 968]|
|:large_blue_circle:|com/starrocks/server/OlapTableFactory.java|5|8|62.50%|[339, 341, 342]|
|:large_blue_circle:|com/starrocks/catalog/TableProperty.java|4|6|66.67%|[386, 387]|
|:large_blue_circle:|com/starrocks/common/util/PeriodStyle.java|40|47|85.11%|[37, 38, 40, 41, 43, 50, 55]|
|:large_blue_circle:|com/starrocks/common/util/DurationStyle.java|38|41|92.68%|[37, 42, 104]|
|:large_blue_circle:|com/starrocks/catalog/PartitionKey.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/OlapScanNode.java|12|12|100.00%|[]|
|:large_blue_circle:|com/starrocks/clone/DynamicPartitionScheduler.java|2|2|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/common/util/DurationStyle.java [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+            }
+            return Duration.ofSeconds(seconds);
+        }
+
+        @Override
+        public String toString(Duration duration) {
+            if (duration.isZero()) {
+                return "0 second";
+            }
+            long seconds = duration.getSeconds();
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/common/util/DurationStyle.java:92]: seconds may be bigger than SECONDS_PER_DAY?

File: fe/fe-core/src/main/java/com/starrocks/common/util/TimeUtils.java [hunk 8]
```diff
@@ -335,4 +337,23 @@ public static long getNextValidTimeSecond(long startTimeSecond, long targetTimeS
         long step = difference / intervalSecond + 1;
         return startTimeSecond + step * intervalSecond;
     }
+
+    public static PeriodDuration parseHumanReadablePeriodOrDuration(String text) {
+        try {
+            return PeriodDuration.of(PeriodStyle.LONG.parse(text));
+        } catch (DateTimeParseException ignored) {
+            return PeriodDuration.of(DurationStyle.LONG.parse(text));
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/common/util/TimeUtils.java:345]: what happens if text is "2 years 1 hour"?

File: fe/fe-core/src/main/java/com/starrocks/lake/LakeTable.java [hunk 9]
```diff
@@ -201,4 +210,48 @@ public String getComment() {
...
+        PeriodDuration cacheDuration = getTableProperty().getDataCachePartitionDuration();
+        if (cacheDuration != null && getPartitionInfo().isRangePartition()) {
+            RangePartitionInfo rangePartitionInfo = (RangePartitionInfo) getPartitionInfo();
+            Range<PartitionKey> partitionRange = rangePartitionInfo.getIdToRange(false).get(partition.getId());
+            Range<PartitionKey> dataCacheRange;
+            if (rangePartitionInfo.isPartitionedBy(PrimitiveType.DATETIME)) {
+                LocalDateTime upper = LocalDateTime.now();
+                LocalDateTime lower = upper.minus(cacheDuration);
+                dataCacheRange = Range.closed(PartitionKey.ofDateTime(lower), PartitionKey.ofDateTime(upper));
+                return partitionRange.isConnected(dataCacheRange);
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/lake/LakeTable.java:243]: what will happen to the future partition?
such as the current time is "2023-05-23" and partition range is [2023-06-01, 2023-07-01)

File: fe/fe-core/src/main/java/com/starrocks/catalog/PartitionKey.java [hunk 10]
```diff
@@ -132,6 +135,20 @@ public static PartitionKey createPartitionKey(List<PartitionValue> keys, List<Co
...
+        PartitionKey partitionKey = new PartitionKey();
+        partitionKey.keys.add(new DateLiteral(dateTime, Type.DATETIME));
+        partitionKey.types.add(PrimitiveType.DATETIME);
+        return partitionKey;
+    }
+
+    public static PartitionKey ofDate(LocalDate date) throws AnalysisException {
+        PartitionKey partitionKey = new PartitionKey();
+        partitionKey.keys.add(new DateLiteral(LocalDateTime.of(date, LocalTime.MIN), Type.DATE));
```
Comment (Astralidea) [inline fe/fe-core/src/main/java/com/starrocks/catalog/PartitionKey.java:147]: why here use LocalTime.MIN?

File: fe/fe-core/src/main/java/com/starrocks/clone/DynamicPartitionScheduler.java [hunk 11]
```diff
@@ -503,10 +502,8 @@ private ArrayList<DropPartitionClause> getDropPartitionClauseByTTL(OlapTable ola
...
+                    PartitionKey.ofDateTime(LocalDateTime.now()) : PartitionKey.ofDate(LocalDate.now());
```
Comment (Astralidea) [inline fe/fe-core/src/main/java/com/starrocks/clone/DynamicPartitionScheduler.java:506]: if you do not use DATE_FORMATTER_UNIX you may not supported time like '0001-01-01'

File: fe/fe-core/src/main/java/com/starrocks/common/util/DurationStyle.java (hunk 7 above)
Comment (sduzh) [reply fe/fe-core/src/main/java/com/starrocks/common/util/DurationStyle.java:92]: Yes, and if so the number of hours will be greater than 24, which is fine.

File: fe/fe-core/src/main/java/com/starrocks/lake/LakeTable.java (hunk 9 above)
Comment (sduzh) [reply fe/fe-core/src/main/java/com/starrocks/lake/LakeTable.java:243]: Queries on these partitions will not fill local disk cache.

File: fe/fe-core/src/main/java/com/starrocks/common/util/TimeUtils.java (hunk 8 above)
Comment (sduzh) [reply fe/fe-core/src/main/java/com/starrocks/common/util/TimeUtils.java:345]: Will throw exception

File: fe/fe-core/src/main/java/com/starrocks/lake/LakeTable.java [hunk 12]
```diff
@@ -201,4 +210,48 @@ public String getComment() {
...
+        } catch (AnalysisException ignored) {
+            return true;
+        }
+    }
+
+    private boolean isEnableFillDataCacheImpl(Partition partition) throws AnalysisException {
+        PeriodDuration cacheDuration = getTableProperty().getDataCachePartitionDuration();
+        if (cacheDuration != null && getPartitionInfo().isRangePartition()) {
+            RangePartitionInfo rangePartitionInfo = (RangePartitionInfo) getPartitionInfo();
+            Range<PartitionKey> partitionRange = rangePartitionInfo.getIdToRange(false).get(partition.getId());
```
[the remark this answers is not in the document]
Comment (sduzh) [reply fe/fe-core/src/main/java/com/starrocks/lake/LakeTable.java:239]: Is there any problem with temp partition?

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 119 / 173 (68.79%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/common/util/TimeUtils.java|0|9|00.00%|[343, 344, 345, 350, 351, 352, 353, 355, 356]|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|0|6|00.00%|[1976, 1977, 1979, 1980, 1981, 1982]|
|:large_blue_circle:|com/starrocks/catalog/RangePartitionInfo.java|0|1|00.00%|[561]|
|:large_blue_circle:|com/starrocks/sql/ast/PartitionValue.java|0|2|00.00%|[50, 54]|
|:large_blue_circle:|com/starrocks/lake/LakeTable.java|4|19|21.05%|[230, 231, 238, 239, 241, 242, 243, 244, 245, 246, 247, 248, 249, 250, 254]|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|3|7|42.86%|[2576, 2577, 2578, 2579]|
|:large_blue_circle:|com/starrocks/common/util/PropertyAnalyzer.java|3|5|60.00%|[977, 978]|
|:large_blue_circle:|com/starrocks/server/OlapTableFactory.java|5|8|62.50%|[328, 330, 331]|
|:large_blue_circle:|com/starrocks/catalog/TableProperty.java|4|6|66.67%|[396, 397]|
|:large_blue_circle:|com/starrocks/common/util/PeriodStyle.java|40|47|85.11%|[37, 38, 40, 41, 43, 50, 55]|
|:large_blue_circle:|com/starrocks/common/util/DurationStyle.java|38|41|92.68%|[37, 42, 104]|
|:large_blue_circle:|com/starrocks/catalog/PartitionKey.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/OlapScanNode.java|12|12|100.00%|[]|
|:large_blue_circle:|com/starrocks/clone/DynamicPartitionScheduler.java|2|2|100.00%|[]|

**Review by Astralidea — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 119 / 173 (68.79%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/common/util/TimeUtils.java|0|9|00.00%|[343, 344, 345, 350, 351, 352, 353, 355, 356]|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|0|6|00.00%|[1987, 1988, 1990, 1991, 1992, 1993]|
|:large_blue_circle:|com/starrocks/catalog/RangePartitionInfo.java|0|1|00.00%|[561]|
|:large_blue_circle:|com/starrocks/sql/ast/PartitionValue.java|0|2|00.00%|[50, 54]|
|:large_blue_circle:|com/starrocks/lake/LakeTable.java|4|19|21.05%|[230, 231, 238, 239, 241, 242, 243, 244, 245, 246, 247, 248, 249, 250, 254]|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|3|7|42.86%|[2573, 2574, 2575, 2576]|
|:large_blue_circle:|com/starrocks/common/util/PropertyAnalyzer.java|3|5|60.00%|[977, 978]|
|:large_blue_circle:|com/starrocks/server/OlapTableFactory.java|5|8|62.50%|[344, 346, 347]|
|:large_blue_circle:|com/starrocks/catalog/TableProperty.java|4|6|66.67%|[396, 397]|
|:large_blue_circle:|com/starrocks/common/util/PeriodStyle.java|40|47|85.11%|[37, 38, 40, 41, 43, 50, 55]|
|:large_blue_circle:|com/starrocks/common/util/DurationStyle.java|38|41|92.68%|[37, 42, 104]|
|:large_blue_circle:|com/starrocks/catalog/PartitionKey.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/OlapScanNode.java|12|12|100.00%|[]|
|:large_blue_circle:|com/starrocks/clone/DynamicPartitionScheduler.java|2|2|100.00%|[]|

**Review by wyb — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 119 / 173 (68.79%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/common/util/TimeUtils.java|0|9|00.00%|[343, 344, 345, 350, 351, 352, 353, 355, 356]|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|0|6|00.00%|[1987, 1988, 1990, 1991, 1992, 1993]|
|:large_blue_circle:|com/starrocks/catalog/RangePartitionInfo.java|0|1|00.00%|[561]|
|:large_blue_circle:|com/starrocks/sql/ast/PartitionValue.java|0|2|00.00%|[50, 54]|
|:large_blue_circle:|com/starrocks/lake/LakeTable.java|4|19|21.05%|[230, 231, 238, 239, 241, 242, 243, 244, 245, 246, 247, 248, 249, 250, 254]|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|3|7|42.86%|[2573, 2574, 2575, 2576]|
|:large_blue_circle:|com/starrocks/common/util/PropertyAnalyzer.java|3|5|60.00%|[977, 978]|
|:large_blue_circle:|com/starrocks/server/OlapTableFactory.java|5|8|62.50%|[344, 346, 347]|
|:large_blue_circle:|com/starrocks/catalog/TableProperty.java|4|6|66.67%|[396, 397]|
|:large_blue_circle:|com/starrocks/common/util/PeriodStyle.java|40|47|85.11%|[37, 38, 40, 41, 43, 50, 55]|
|:large_blue_circle:|com/starrocks/common/util/DurationStyle.java|38|41|92.68%|[37, 42, 104]|
|:large_blue_circle:|com/starrocks/catalog/PartitionKey.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/OlapScanNode.java|12|12|100.00%|[]|
|:large_blue_circle:|com/starrocks/clone/DynamicPartitionScheduler.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 119 / 173 (68.79%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/common/util/TimeUtils.java|0|9|00.00%|[343, 344, 345, 350, 351, 352, 353, 355, 356]|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|0|6|00.00%|[1979, 1980, 1982, 1983, 1984, 1985]|
|:large_blue_circle:|com/starrocks/catalog/RangePartitionInfo.java|0|1|00.00%|[561]|
|:large_blue_circle:|com/starrocks/sql/ast/PartitionValue.java|0|2|00.00%|[50, 54]|
|:large_blue_circle:|com/starrocks/lake/LakeTable.java|4|19|21.05%|[239, 240, 247, 248, 250, 251, 252, 253, 254, 255, 256, 257, 258, 259, 263]|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|3|7|42.86%|[2570, 2571, 2572, 2573]|
|:large_blue_circle:|com/starrocks/common/util/PropertyAnalyzer.java|3|5|60.00%|[977, 978]|
|:large_blue_circle:|com/starrocks/server/OlapTableFactory.java|5|8|62.50%|[344, 346, 347]|
|:large_blue_circle:|com/starrocks/catalog/TableProperty.java|4|6|66.67%|[396, 397]|
|:large_blue_circle:|com/starrocks/common/util/PeriodStyle.java|40|47|85.11%|[37, 38, 40, 41, 43, 50, 55]|
|:large_blue_circle:|com/starrocks/common/util/DurationStyle.java|38|41|92.68%|[37, 42, 104]|
|:large_blue_circle:|com/starrocks/catalog/PartitionKey.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/OlapScanNode.java|12|12|100.00%|[]|
|:large_blue_circle:|com/starrocks/clone/DynamicPartitionScheduler.java|2|2|100.00%|[]|

**Comment by sduzh** [comment]
@Mergifyio backport branch-3.1

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 119 / 173 (68.79%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/common/util/TimeUtils.java|0|9|00.00%|[343, 344, 345, 350, 351, 352, 353, 355, 356]|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|0|6|00.00%|[1977, 1978, 1980, 1981, 1982, 1983]|
|:large_blue_circle:|com/starrocks/catalog/RangePartitionInfo.java|0|1|00.00%|[561]|
|:large_blue_circle:|com/starrocks/sql/ast/PartitionValue.java|0|2|00.00%|[50, 54]|
|:large_blue_circle:|com/starrocks/lake/LakeTable.java|4|19|21.05%|[239, 240, 247, 248, 250, 251, 252, 253, 254, 255, 256, 257, 258, 259, 263]|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|3|7|42.86%|[2565, 2566, 2567, 2568]|
|:large_blue_circle:|com/starrocks/common/util/PropertyAnalyzer.java|3|5|60.00%|[977, 978]|
|:large_blue_circle:|com/starrocks/server/OlapTableFactory.java|5|8|62.50%|[344, 346, 347]|
|:large_blue_circle:|com/starrocks/catalog/TableProperty.java|4|6|66.67%|[396, 397]|
|:large_blue_circle:|com/starrocks/common/util/PeriodStyle.java|40|47|85.11%|[37, 38, 40, 41, 43, 50, 55]|
|:large_blue_circle:|com/starrocks/common/util/DurationStyle.java|38|41|92.68%|[37, 42, 104]|
|:large_blue_circle:|com/starrocks/catalog/PartitionKey.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/OlapScanNode.java|12|12|100.00%|[]|
|:large_blue_circle:|com/starrocks/clone/DynamicPartitionScheduler.java|2|2|100.00%|[]|

**Review by wyb — APPROVED** [review]

**Review by gengjun-git — APPROVED** [review]

**Pull request merged**